File: 3490-find-the-maximum-length-of-valid-subsequence-i/find-the-maximum-length-of-valid-subsequence-i.py
class Solution:
    def maximumLength(self, nums: List[int]) -> int:
        evens = list(filter(lambda x: x % 2 == 0, nums))
        odds = list(filter(lambda x: x % 2 != 0, nums))

        stack = []

        for num in nums:
            if not stack:
                stack.append(num)
                continue
            
            if stack[-1] % 2 == 0 and num % 2 != 0:
                stack.append(num)

            if stack[-1] % 2 != 0 and num % 2 == 0:
                stack.append(num)

        print(evens)
        print(odds)
        print(stack)
        return max(
            len(evens), len(odds), len(stack)
        )
        

File: 3490-find-the-maximum-length-of-valid-subsequence-i/find-the-maximum-length-of-valid-subsequence-i.rs
use std::cmp;

impl Solution {
    pub fn maximum_length(nums: Vec<i32>) -> i32 {
        let mut even = 0;
        let mut odd = 0;
        let mut assemble = 0;
        let mut assemble_last = 0;

        for num in &nums {
            if num % 2 == 0 {
                even += 1;
                
                if assemble == 0 {
                    assemble += 1;
                    assemble_last = *num;
                    continue;
                } 

                if assemble_last % 2 != 0 {
                    assemble += 1;
                    assemble_last = *num;
                }
  
            } else {
                odd += 1;
                if assemble == 0 {
                    assemble += 1;
                    assemble_last = *num;
                    continue;
                } 

                if assemble_last % 2 == 0 {
                    assemble += 1;
                    assemble_last = *num;
                }
            }
        }

        cmp::max(cmp::max(even, odd), assemble) 
    }
}
